skops/io/_dispatch.py

```diff
@@ -2,34 +2,233 @@
+    audit methods are called after ``__init__`` and before ``construct``.
+    Therefore ``__init__`` should avoid creating any instances or importing
+    any modules, to avoid running unwanted code.
+
+    Parameters
+    ----------
+    state : dict
+        A dict representing the state of the dumped object.
+
+    load_context : LoadContext
+        The context of the loading process.
+
+    trusted : bool or list, default=False
+        If ``True``, the object will be loaded without any security checks. If
+        ``False``, the object will be loaded only if there are only trusted
+        objects in the dumped file. If a list of strings, the object will be
+        loaded only if there are only trusted objects and objects of types
+        listed in ``trusted`` are in the dumped file.
+    """
+
+    def __init__(self, state, load_context: LoadContext, trusted=False):
+        self.class_name, self.module_name = state["__class__"], state["__module__"]
+        self.trusted = trusted
+        self._is_safe = None
+        self._constructed = None
+
+    def construct(self):
+        """Construct the object.
+
+        We only construct the object once, and then cache the result.
+        """
+        if self._constructed is not None:
+            return self._constructed
+        self._constructed = self._construct()
+        return self._constructed
+
+    @classmethod
+    def _get_trusted(cls, trusted, default):
+        """Return a trusted list, or True.
+
+        If `trusted` is `False`, we return the `defaults`, otherwise the
+        `trusted` value is used.
+
+        This is a convenience method called by child classes.
+        """
+        if trusted is True:
+            # if trusted is True, we trust the node
+            return True
+        elif trusted is False:
+            # if trusted is False, we only trust the defaults
+            return default
+        # otherwise we trust the given list
+        return trusted
+
+    def _get_iterable_safety(self, values):
+        """Check if members of an iterable are all safe."""
+        for item in values:
+            if not item.is_safe:
+                return False
+        return True
+
+    @property
+    def is_self_safe(self):
+        """True only if the node's type is considered safe.
+
+        This property only checks the type of the node, not its children.
+        """
+        return check_type(self.module_name, self.class_name, self.trusted)
+
+    @property
+    def is_safe(self):
+        """Trie only if the node and all its children are safe."""
+        # the safety value is cached.
+        if self._is_safe is not None:
+            return self._is_safe
+
+        # if trusted is set to True, we don't do any safety checks.
+        if self.trusted is True:
+            self._is_safe = True
+            return True
+
+        is_safe = self.is_self_safe
+
+        for child, _type in self.children.items():
```

Comment:
I wonder if we should make `children` a list of tuples instead of a dict, where the first element is the attribute itself and not its name. The reason why is that if I'm not mistaken, we always do `getattr(self, child)` (where `child` is the name), so would it not be better to store the attribute itself instead of `getattr`ing it with its name? Besides performance, this also leaves a bit less room for error, e.g. that I have a typo in the name when setting `children`. Obviously, we have to use a list of tuples instead of a dict for this to work, but there is no strict reason why it needs to be a dict, right?